nomad/vault.go

```diff
@@ -465,17 +465,18 @@ OUTER:
 		case <-v.tomb.Dying():
 			return
 		case <-retryTimer.C:
-			// Ensure the API is reachable
-			if !initStatus {
-				if _, err := v.clientSys.Sys().InitStatus(); err != nil {
-					v.logger.Warn("failed to contact Vault API", "retry", v.config.ConnectionRetryIntv, "error", err)
-					retryTimer.Reset(v.config.ConnectionRetryIntv)
-					continue OUTER
-				}
-				initStatus = true
-			}
 			// Retry validating the token till success
 			if err := v.parseSelfToken(); err != nil {
+				// if parsing token fails, try to distinguish legitimate token error from transient Vault initialization/connection issue
+				if !initStatus {
+					if _, err := v.clientSys.Sys().Health(); err != nil {
+						v.logger.Warn("failed to contact Vault API", "retry", v.config.ConnectionRetryIntv, "error", err)
+						retryTimer.Reset(v.config.ConnectionRetryIntv)
+						continue OUTER
+					}
+					initStatus = true
+				}
```

Comment:
@schmichael Can I get a second review please?  I have restored the health check  because of the way CreateToken handles connection errors:

https://github.com/hashicorp/nomad/blob/a3b4f060fb49ed2c6408ffce2fd01454d7fdd209/nomad/vault.go#L951-L956

In the above code, we aim to detect if the server failed to make any API contact, vault will return a recoverable error, but if /sys/init (or health now) succeeds, then it marks the token failures as permenant.

I don't buy into this distinction much, but to be conservative in a point release, I brought back the health check.  However, I made reversed the order of the check - only check health if parseToken failed.  It should be functionally the same as before.

Reply:
Should we instead make errors from parseSelfToken recoverable?

Reply:
I don't buy into the distinction but I'm afraid of making of any unintended side-effects considering it seems it was explicitly designed this way and this file has bitten as before.  I'm more inclined to keep the change conservative for 0.12.X and make more ambitious refactor in a major release.